plugins/channel-safely/channel-safely-plugin.cpp

```diff
@@ -0,0 +1,211 @@
+    if (parameters.empty()) {
+        if (debug_out) debug_out->print("mcd->manage_designations()\n");
+        ChannelManager::Get().manage_designations(out);
+        if (!enabled) {
+            ChannelManager::Get().delete_groups();
+        }
+    }
+    // enable options
+    else if (parameters.size() == 1 && parameters[0] == "enable") {
+        return plugin_enable(out, true);
+    } else if (parameters.size() == 2 && parameters[0] == "enable" && parameters[1] == "cheats" && !cheat_mode) {
+        cheat_mode = true;
+        out.print("channel-safely: cheat mode enabled!\n");
+    }
+    // disable options
+    else if (parameters.size() == 1 && parameters[0] == "disable") {
+        return plugin_enable(out, false);
+    } else if (parameters.size() == 2 && parameters[0] == "disable" && parameters[1] == "cheats" && cheat_mode) {
+        cheat_mode = false;
+        out.print("channel-safely: ontick events disabled!\n");
+    }
+    // developer debug
+    else if (parameters.size() == 1 && parameters[0] == "debug") {
+        debug_out = &out;
+        ChannelManager::Get().debug();
+        debug_out = nullptr;
+    } else if (parameters.size() == 1 && parameters[0] == "help") {
+        return CR_WRONG_USAGE;
+    } else {
+        debug_out = nullptr;
+        out.color(color_value::COLOR_RED);
+        out.print("Invalid argument.\n\n");
+        out.color(color_value::COLOR_RESET);
+        return CR_WRONG_USAGE;
+    }
+    debug_out = nullptr;
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    namespace EM = EventManager;
+#ifdef CS_DEBUG
+    debug_out = &out;
+    out.print("debugging enabled\n");
+#endif
+    if (enable && !enabled) {
+        using namespace EM::EventType;
+        EM::EventHandler jobStartHandler(onJobStart, 0);
+        EM::EventHandler jobCompletionHandler(onJobComplete, 0);
+        EM::registerListener(EventType::JOB_INITIATED, jobStartHandler, plugin_self);
+        EM::registerListener(EventType::JOB_COMPLETED, jobCompletionHandler, plugin_self);
+        ChannelManager::Get().manage_designations(out);
+        out.print("channel-safely: enabled!\n");
+    } else if (!enable) {
+        ChannelManager::Get().delete_groups();
+        EM::unregisterAll(plugin_self);
+        out.print("channel-safely: disabled!\n");
+    }
+    enabled = enable;
+    debug_out = nullptr;
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+#ifdef CS_DEBUG
+    debug_out = &out;
+#endif
+    if (debug_out) debug_out->print("onstatechange()\n");
+    if (enabled && World::isFortressMode() && Maps::IsValid()) {
+        switch (event) {
+            case SC_MAP_LOADED:
+                if (debug_out) debug_out->print("SC_MAP_LOADED\n");
+                ChannelManager::Get().manage_designations(out);
+                break;
+            case SC_PAUSED:
```

Comment:
why are we running on pause events? what state could have changed from the last handled event?

Reply:
They may pause to designate an entire fortress. That's why I run on unpause. I run on pause, simply because that's a good time to run (the game is doing nothing so there are plenty of cpu cycles to go around). I can likely remove some more triggers, but I went with `better safe than sorry`

Reply:
Unpause I totally agree with. Even theoretically, though, what new data could be found by the pause trigger? If nothing new was found by the now-removed tick events, why are we keeping pause?

Reply:
I think actually that tick event was necessary. I forgot that the job events don't build groups, the only check/update with the job they have.

I'm going to finish going through and commenting.. but for now I think I'll be adding that tick event back (wups).. or maybe more likely once the commenting is done we have a discussion on what makes most sense.

Reply:
Ok, so finished adding inline/function comments; just remembered I left the class comments unfinished (oops), they're less important though I think. Reorganized the source, separating the classes. It means more files, but I think it's easier to find things now.

So.. I do think the tick event was performing a critical operation in that it prevented the job events from lacking group data.. but with a 1 day frequency it certainly could have missed something in between.

I added
```cpp
void ChannelManager::manage_one(color_ostream &out, const df::coord &map_pos, df::map_block* block, bool manage_neighbours) {
    if (!groups.count(map_pos)) {
        build_groups();
    }
```
which get's called by the events, so that should avoid some problems.. but it might miss new designations above it.. now that I think about that..

But.. Now that's got a solution too.. I added a build call inside `has_groups_above` which also now checks if there is an untracked designation above it..

So with all this I think it makes 100% sense to remove the Pause trigger